ouroboros/chainsync/types.go

```diff
@@ -387,13 +789,268 @@ type Tip struct {
 	Height uint64 `json:"height,omitempty" dynamodbav:"height,omitempty"`
 }
 
+func (t Tip) String() string {
+	return fmt.Sprintf("slot=%v id=%v height=%v", t.Slot, t.ID, t.Height)
+}
+
+type TipV5 struct {
+	Slot    uint64 `json:"slot,omitempty"    dynamodbav:"slot,omitempty"`
+	Hash    string `json:"hash,omitempty"    dynamodbav:"hash,omitempty"`
+	BlockNo uint64 `json:"blockNo,omitempty" dynamodbav:"blockNo,omitempty"`
+}
+
+// Support findIntersect (v6) / FindIntersection (v5) universally.
+type CompatibleResponsePraos ResponsePraos
+
+func (c *CompatibleResponsePraos) UnmarshalJSON(data []byte) error {
+	var r ResponsePraos
+	err := json.Unmarshal(data, &r)
+	if err == nil && r.Result != nil {
+		*c = CompatibleResponsePraos(r)
+		return nil
+	}
+
+	var r5 ResponseV5
+	err = json.Unmarshal(data, &r5)
+	c.JsonRpc = "2.0"
+	if err != nil {
+		// Just skip all the data processing, as it's useless.
+		return nil
+	} else {
+		// All we really care about is the result.
+		if r5.Result.IntersectionFound != nil {
+			c.Method = FindIntersectionMethod
+
+			var p Point
+			p.pointType = r5.Result.IntersectionFound.Point.pointType
+			if p.pointType == PointTypeString {
+				p.pointString = r5.Result.IntersectionFound.Point.pointString
+			} else if p.pointType == PointTypeStruct {
+				p.pointStruct.Slot = r5.Result.IntersectionFound.Point.pointStruct.Slot
+				p.pointStruct.ID = r5.Result.IntersectionFound.Point.pointStruct.Hash
+			} else {
+				panic("Invalid point type")
+			}
+
+			var t Tip
+			t.Slot = r5.Result.IntersectionFound.Tip.Slot
+			t.ID = r5.Result.IntersectionFound.Tip.Hash
+			t.Height = r5.Result.IntersectionFound.Tip.BlockNo
+
+			var findIntersection CompatibleResultFindIntersection
+			findIntersection.Intersection = &p
+			findIntersection.Tip = &t
+			c.Result = &findIntersection
+		} else if r5.Result.IntersectionNotFound != nil {
+			c.Method = FindIntersectionMethod
+
+			var t Tip
+			t.Slot = r5.Result.IntersectionNotFound.Tip.Slot
+			t.ID = r5.Result.IntersectionNotFound.Tip.Hash
+			t.Height = r5.Result.IntersectionFound.Tip.BlockNo
+
+			var e ResultError
+			e.Data = &t
+			e.Code = 1000
+			e.Message = "Intersection not found - Conversion from a v5 Ogmigo call"
+			c.Error = &e
+		} else if r5.Result.RollForward != nil {
+			c.Method = NextBlockMethod
```

Comment:
This looks like a lot of duplication with the above code; what's going on here?

Reply:
Ahhh, I forgot to write some helper functions that cut down on duplication. I fixed that.